File: src/com/pat/task05/Main.java
package com.pat.task05;

import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        try {
            while (in.hasNextInt()) {
                int K = in.nextInt();
                if (K < 1 || K >= 100) return;
                List<Integer> arr = new ArrayList<>();
                List<Integer> res = new ArrayList<>();
                for (int i = 0; i < K; i++) {
                    int tmp = in.nextInt();
                    if (!(tmp > 1 && tmp <= 100)) return;
                    arr.add(tmp);
                }
                for (int i = 0; i < K; i++) {
                    if ((arr.get(i) * 2 - 1) % 3 == 0) {
                        if (arr.indexOf(arr.get(i) * 2) == -1 && arr.indexOf((arr.get(i) * 2 - 1) / 3) == -1) {
                            res.add(arr.get(i));
                        }
                    } else if (arr.indexOf(arr.get(i) * 2) == -1) {
                        res.add(arr.get(i));
                    }
                }
                Collections.sort(res);
                Collections.reverse(res);
                for (int i = 0; i < res.size(); i++) {
                    System.out.print(res.get(i));
                    if (i != res.size() - 1) {
                        System.out.print(" ");
                    }
                }
            }
        } catch (NullPointerException e) {
            e.printStackTrace();
        }
    }
}

File: src/com/pat/task05/main.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using std::cin;
using std::cout;
using std::endl;
using std::vector;

int main()
{
    int K = 0;
    cin >> K;
    vector<int> arr;
    int flagArr[10000] = { 0 };
    for (int i = 0; i < K; ++i)
    {
        int tmp;
        cin >> tmp;
        arr.push_back(tmp);
        while (tmp != 1)
            if ((tmp & 1) == 0)
                flagArr[tmp /= 2] = 1;
            else
                flagArr[tmp = (tmp * 3 + 1) / 2] = 1;
    }

    std::sort(arr.begin(), arr.end(), std::greater<int>());

    bool hasPrinted = false;
    for (int i = 0; i < K; ++i)
    {
        if (flagArr[arr[i]] == 0)
        {
            if (hasPrinted) cout << " " << arr[i];
            else
            {
                cout << arr[i];
                hasPrinted = true;
            }
        }
    }

    return 0;
}
